pkg/controller/velerobackup/hiveobject.go

```diff
@@ -0,0 +1,113 @@
+package velerobackup
+
+import (
+	"fmt"
+
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+	log "github.com/sirupsen/logrus"
+	meta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+)
+
+const (
+	errChecksum = "HIVE_CHECKSUM_ERR_97A29D08"
+)
+
+type hiveObject struct {
+	checksum string
+	object   runtime.Object
+	logger   log.FieldLogger
+}
+
+func newHiveObject(object runtime.Object, logger log.FieldLogger) (*hiveObject, error) {
+	switch t := object.(type) {
+	case *hivev1.ClusterDeployment:
+		checksum := calculateChecksum(t.ObjectMeta, t.Spec, controllerutils.GetChecksumOfObjects, logger)
+		return &hiveObject{
+			checksum: checksum,
+			object:   object,
+			logger:   logger,
+		}, nil
+	case *hivev1.SyncSet:
+		checksum := calculateChecksum(t.ObjectMeta, t.Spec, controllerutils.GetChecksumOfObjects, logger)
+		return &hiveObject{
+			checksum: checksum,
+			object:   object,
+			logger:   logger,
+		}, nil
+	case nil:
+		return nil, fmt.Errorf("Can't convert nil to a HiveObject")
+	}
+
+	return nil, fmt.Errorf("Unknown Type: %T", object)
+}
+
+func (h *hiveObject) hasChanged() bool {
+	if h.checksum == errChecksum {
+		// The checksum function errored. In this case we assume the object has NOT changed.
+		return false
+	}
+
+	// Make sure newChecksum is different than the old checksum.
+	oldChecksum := h.getChecksumAnnotation()
+	return h.checksum != oldChecksum
+}
+
+func calculateChecksum(meta metav1.ObjectMeta, spec interface{}, checksumFunc controllerutils.ChecksumOfObjectsFunc, logger log.FieldLogger) string {
+	// We need to take the checksum WITHOUT the previous checksum
+	// to avoid an infinite loop of the checksum always changing.
+	metaCopy := meta.DeepCopy()
+	delete(metaCopy.Annotations, controllerutils.LastBackupAnnotation)
+
+	// We also need to take the checksum WITHOUT the previous ResourceVersion
+	// as when we update the checksum annotation, ResourceVersion changes.
+	metaCopy.ResourceVersion = ""
+
+	checksum, err := checksumFunc(metaCopy, spec)
+	if err != nil {
+		// We never return an error from this function as that could lead to a
+		// situation where other objects change, but because one of the object
+		// checksum errored, then the other objects won't get backed up.
+		// Instead we return a constant checksum value. This function should rarely,
+		// if ever, return an error.
+		logger.WithError(err).Info("error calculating object checksum")
+		return errChecksum
+	}
+
+	return checksum
+}
+
+func (h *hiveObject) getChecksumAnnotation() string {
+	tmpmeta, err := meta.Accessor(h.object)
+	if err != nil {
+		// This should never happen because we should always be an object.
+		panic(err)
```

Comment:
Understanding that this is a logic error and not a runtime error, we still should not panic. Instead, log it as an error and return an empty string, treating it as though the annotation is not set.